openmdao/components/meta_model_semi_structured_comp.py

```diff
@@ -118,11 +118,20 @@ def add_output(self, name, training_data, **kwargs):
         n = self.options['vec_size']
         super().add_output(name, np.ones(n), **kwargs)
 
-        self.training_outputs[name] = training_data
-
         if self.options['training_data_gradients']:
+            if training_data is None:
+                for item in self.training_inputs.values():
+                    n_train = len(item)
+                    break
```

Comment:
maybe a code comment here?  
in isolation, it's not clear why there is an iterator with an immediate break...